IBX-6404: Added create bundle extension (#2214)
ibexa/documentation-developer: docs/administration/project_organization/bundle_extension/package_structure.md

# Package and bundle structure and namespaces

Bundle allows you to extend the system by adding new services, routes, controllers, and views, and extend functionality to fully customize your [[= product_name =]] application.
If you want to contribute to [[= product_name =]] development, you need to adhere to the package and bundle structure and namespace standards.

All conventions described here apply to contributions to [[= product_name_base =]] core code, not to third party packages.

!!! note
    New code needs to follow the rules outlined here.
    They are being applied progressively to existing code.

> 
> 
> 

Reviewer: ```suggestion

```

## Root PHP namespace

Define [[= product_name =]] core PHP code in a namespace with the following prefix:

> 
> 
> 

Reviewer: ```suggestion

```

```php
namespace Ibexa;
```

A package which groups some DXP features can use an additional prefix, for example:

> 
> 
> 

Reviewer: ```suggestion

```

```php
namespace Ibexa\Commerce;

> ```
> 
> 
> 

Reviewer: ```suggestion
```

```

```php
namespace Ibexa\Personalization;
```

> 
> 

Reviewer: ```suggestion
```

## Packages

The general package directory structure and corresponding PHP namespace mappings are:

> 
> 
> 

Reviewer: ```suggestion

```

```
.
+-- src
|   +-- bundle (`Ibexa\Bundle\<PackageName>`)
|   +-- contracts (`Ibexa\Contracts\<PackageName>`)
|   +-- lib (`Ibexa\<PackageName>`)
+-- tests
|   +-- bundle (`Ibexa\Tests\Bundle\<PackageName>`)
|   +-- integration (`Ibexa\Tests\Integration\<PackageName>`)
|   +-- lib (`Ibexa\Tests\<PackageName>`)
```

If a package doesn't contain some of the described parts, you can skip those directories.

> 
> 

Reviewer: ```suggestion
```

### Implementation (lib)

The `src/lib` directory and its corresponding `Ibexa\<PackageName>` namespace are meant for internal implementation not tied to the Symfony Framework.

Examples:

> 
> 

Reviewer: ```suggestion
```

```php
namespace Ibexa\Search;
```

> 
> 

Reviewer: ```suggestion
```

```php
namespace Ibexa\Commerce\Shop;
```

> 
> 

Reviewer: ```suggestion
```

### Bundles

> The bundle class definition in the `src/bundle` directory must have the following pattern:

Reviewer: ```suggestion
The bundle class definition in the `src/bundle` directory must follow the pattern:
```

> 
> 

Reviewer: ```suggestion
```

```php
namespace Ibexa\Bundle\<PackageName>;
 
class Ibexa[ProductGroup]<PackageName>Bundle // ...
```

Examples:

```php
namespace Ibexa\Bundle\Search;
 
class IbexaSearchBundle // ...
```

> 
> 

Reviewer: ```suggestion
```

```php
namespace Ibexa\Bundle\Commerce\Shop;
class IbexaCommerceShopBundle // ...
```

> 
> 

Reviewer: ```suggestion
```

### Contracts

A package may introduce a namespace for contracts, to be consumed by first and third party packages
and projects, which must be prefixed as:

> 
> 

Reviewer: ```suggestion
```

```php
namespace Ibexa\Contracts;
```

Examples:

> 
> 

Reviewer: ```suggestion
```

```php
namespace Ibexa\Contracts\Kernel;
```

> 
> 

Reviewer: ```suggestion
```

```php
namespace Ibexa\Contracts\SiteFactory;
```

> 
> 

Reviewer: ```suggestion
```

# Package and bundle structure and namespaces

> Bundle allows you to extend the system by adding new services, routes, controllers, and views, and extend functionality to fully customize your [[= product_name =]] application.

Reviewer 2: ```suggestion
Bundles allow you to extend the system by adding new services, routes, controllers, and views, and fully customize your [[= product_name =]] application by extending its functionality.
```

If you want to contribute to [[= product_name =]] development, you need to adhere to the package and bundle structure and namespace standards.

All conventions described here apply to contributions to [[= product_name_base =]] core code, not to third party packages.

!!! note
    New code needs to follow the rules outlined here.
    They are being applied progressively to existing code.
## Root PHP namespace

Define [[= product_name =]] core PHP code in a namespace with the following prefix:
```php
namespace Ibexa;
```

A package which groups some DXP features can use an additional prefix, for example:
```php
namespace Ibexa\Commerce;
```php
namespace Ibexa\Personalization;
```

## Packages

> The general package directory structure and corresponding PHP namespace mappings are:

Reviewer 2: ```suggestion
The general package directory structure and corresponding PHP namespace mappings are as follows:
```
This structure uses different formatting than the one in the create bundle article. Could we decide on one and use it consistently?
